src/hotspot/share/code/nmethod.cpp

```diff
@@ -1486,26 +1499,24 @@ nmethod::nmethod(
     } else {
       _unwind_handler_offset = -1;
     }
-    CHECKED_CAST(_metadata_offset, uint16_t, (align_up(code_buffer->total_oop_size(), oopSize)));
-    int metadata_end_offset = _metadata_offset + align_up(code_buffer->total_metadata_size(), wordSize);
 
-#if INCLUDE_JVMCI
-    CHECKED_CAST(_jvmci_data_offset, uint16_t, metadata_end_offset);
-    int jvmci_data_size   = compiler->is_jvmci() ? jvmci_data->size() : 0;
-    DEBUG_ONLY( int data_end_offset = _jvmci_data_offset  + align_up(jvmci_data_size, oopSize); )
-#else
-    DEBUG_ONLY( int data_end_offset = metadata_end_offset; )
-#endif
-    assert((data_offset() + data_end_offset) <= nmethod_size, "wrong nmethod's size: %d > %d",
-           (data_offset() + data_end_offset), nmethod_size);
+    CHECKED_CAST(_oops_size, uint16_t, align_up(code_buffer->total_oop_size(), oopSize));
+    uint16_t metadata_size = (uint16_t)align_up(code_buffer->total_metadata_size(), wordSize);
+    JVMCI_ONLY(CHECKED_CAST(_jvmci_data_size, uint16_t, align_up(compiler->is_jvmci() ? jvmci_data->size() : 0, oopSize)));
```

Comment:
This cast is lossy in that `jvmci_data->size()` returns an int. It caused a `double free or corruption (out)` crash in Graal in the case where a `JVMCINMethodData` had a very long name. We've fixed this by [limiting the length of the name](https://github.com/openjdk/jdk/pull/24753) but I'm wondering if there was some special reason for this cast?

Reply:
The cast was added by 8331087, which reduced the supported JVMCI data size to uint16_t.  I don't remember this issue with long names coming up during that review, so I guess we all missed it.  @dougxc please file a bug so we can track this.  It seems like JVMCINMethodData::copy should do something like truncate long names rather than blindly assuming it has enough space.  If uint16_t is unreasonably small for JVMCI nmethod data we could revert that change and make it 32 bits again.

Reply:
Yes, my mistake.  I was thinking `_jvmci_data_offset` was used to compute `jvmci_data_end()`, not `jvmci_data_begin()`.

Reply:
Yes we should use 32 bits. Even if we revert back to using _jvmci_data_offset we can use uint16_t because size of relocation (after which JVMCI data is placed) data is bigger.

Reply:
Thanks for the report. Yes, cast to uint16 is wrong. I am going to fix the issue here: https://github.com/openjdk/jdk/pull/24965